hashes/src/sha256t.rs

```diff
@@ -7,14 +7,39 @@ use core::marker::PhantomData;
 
 use crate::{sha256, FromSliceError, HashEngine as _};
 
-type HashEngine = sha256::HashEngine;
+/// Engine to compute SHA256t hash function.
+#[derive(Clone)]
+pub struct HashEngine(sha256::HashEngine);
```

Comment:
I intended this to be generic over tag, so people also don't mix up different tags. It is also required to fix the `Default` problem. It might bring new challenges to macros or other code, so if you try it and find a problem don't hesitate to ask me for help. I was actually going to do this if you didn't beat me to it.

Reply:
>  It is also required to fix the Default problem. 

Since @apoelstra thumbs upped then I guess I'm missing something but I don't think it is.

Reply:
@tcharding we want to have `Default` on the tagged hash engines, just like we have it for the other engine types. This PR simply removes that (by creating a newtype that doesn't have it). If we made the engine generic over the tag type, then we could have it.

Normally I'd say "this PR is fine as-is, let's add `Default` in a followup" .. but since you're creating a new type that we want to change to be generic, we should probably just do it in one go.

Reply:
> I intended this to be generic over tag, so people also don't mix up different tags.

This is a totally different problem to the one I am solving in this PR - which is explicitly the footgun of accidentally using a `sha256::HashEngine::default()` instead of the one returned by `Tag`s `engine` method.

Reply:
Oh I remember now, I personally think the `Default` problem we have is that `GeneralHash` should not have a default implementation of `engine` (which calls through to `Default`) - I haven't got a working patch set though so this discussion is just to share whats in my head. Will push up a PR for you guys to look at. No need to comment because I need to convince myself anyways if its wrong.

Reply:
> This is a totally different problem to the one I am solving in this PR - which is explicitly the footgun of accidentally using a `sha256::HashEngine::default()` instead of the one returned by `Tag`s `engine` method.

It's not a totally different problem. The problem you're describing is because `Tag::engine` returns a weakly-typed object. It should return a strongly-typed one which cannot be confused with the engine for other hashes.

Reply:
> `GeneralHash` should not have a default implementation of `engine` (which calls through to `Default`)

There's no reason it shouldn't if the types for different hashes are distinct. FWIW I consider tagged hashes distinct from each other (as if they were completely different hash functions).